src/main/java/org/opensearch/sdk/BaseExtension.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.sdk;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import org.apache.lucene.util.SetOnce;
+import org.opensearch.client.Client;
+import org.opensearch.client.node.NodeClient;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.io.stream.NamedWriteableRegistry;
+import org.opensearch.common.io.stream.NamedWriteableRegistry.Entry;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.settings.Setting.Property;
+import org.opensearch.common.xcontent.NamedXContentRegistry;
+import org.opensearch.env.Environment;
+import org.opensearch.env.NodeEnvironment;
+import org.opensearch.node.Node;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.script.ScriptContext;
+import org.opensearch.script.ScriptEngine;
+import org.opensearch.script.ScriptService;
+import org.opensearch.threadpool.ExecutorBuilder;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.watcher.ResourceWatcherService;
+
+public abstract class BaseExtension implements Extension {
+    private Client client;
+    private ClusterService clusterService;
+    private ThreadPool threadPool;
+    private ResourceWatcherService resourceWatcherService;
+    private ScriptService scriptService;
+    private NamedXContentRegistry xContentRegistry;
+    private Environment environment;
+    private NodeEnvironment nodeEnvironment;
+    private NamedWriteableRegistry namedWriteableRegistry;
+    private IndexNameExpressionResolver indexNameExpressionResolver;
+    private Supplier<RepositoriesService> repositoriesServiceSupplier;
+
+    /**
+     * 
+     */
+    public BaseExtension() throws IOException {
```

Comment:
I'm really kind of iffy on just passing through an exception in a constructor.   Would prefer a nice and verbose error message to the user wrapping (caused by) the thrown exception.  And since this is going to be fatal to the Extension startup, I think an `Error` is more appropriate than an `Exception`.  This also allows us to remove the `throws` clause.

Recommend a try/catch around create components, and in the catch block throw an `IOError` with a message: "Extension failed to initialize due to an exception creating a component: " + exception.getMessage(), and then linking to the exception so the stack trace provides more info.

Open to others' ideas on this as well.